File: services/abdm/carecontext/webhook_test.go
package carecontext

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"testing"
	"time"
)

func sign(t *testing.T, body, secret string, ts time.Time) string {
	t.Helper()
	sec := fmt.Sprintf("%d", ts.Unix())
	m := hmac.New(sha256.New, []byte(secret))
	m.Write([]byte(sec + "." + body))
	return fmt.Sprintf("t=%s,v1=%s", sec, hex.EncodeToString(m.Sum(nil)))
}

const secret = "whsec"

func TestParseWebhookSignature(t *testing.T) {
	body := `{"service":"abdm","event":"abha.link_care_context","data":{"care_context_id":"cc-1","status":"LINKED"}}`

	t.Run("valid", func(t *testing.T) {
		e, err := ParseWebhook([]byte(body), sign(t, body, secret, time.Now()), secret)
		if err != nil {
			t.Fatalf("ParseWebhook: %v", err)
		}
		st, ok := e.(*LinkStatusEvent)
		if !ok {
			t.Fatalf("event type = %T, want *LinkStatusEvent", e)
		}
		if st.CareContextID != "cc-1" || st.Status != "LINKED" {
			t.Errorf("event = %+v", st)
		}
	})

	t.Run("tampered body", func(t *testing.T) {
		sig := sign(t, body, secret, time.Now())
		_, err := ParseWebhook([]byte(body+" "), sig, secret)
		if !errors.Is(err, ErrBadSignature) {
			t.Errorf("err = %v, want ErrBadSignature", err)
		}
	})

	t.Run("wrong secret", func(t *testing.T) {
		_, err := ParseWebhook([]byte(body), sign(t, body, "other", time.Now()), secret)
		if !errors.Is(err, ErrBadSignature) {
			t.Errorf("err = %v, want ErrBadSignature", err)
		}
	})

	t.Run("stale timestamp", func(t *testing.T) {
		old := time.Now().Add(-10 * time.Minute)
		_, err := ParseWebhook([]byte(body), sign(t, body, secret, old), secret)
		if !errors.Is(err, ErrStaleTimestamp) {
			t.Errorf("err = %v, want ErrStaleTimestamp", err)
		}
	})

	t.Run("malformed header", func(t *testing.T) {
		_, err := ParseWebhook([]byte(body), "garbage", secret)
		if !errors.Is(err, ErrBadSignature) {
			t.Errorf("err = %v, want ErrBadSignature", err)
		}
	})
}

// The upstream event names are inconsistent (note the third lacks the care_
// prefix), so each is asserted verbatim.
func TestParseWebhookEventNames(t *testing.T) {
	for _, tc := range []struct {
		event string
		want  string
	}{
		{"abha.hip_data_fetch", "*carecontext.DataFetchEvent"},
		{"abha.link_care_context", "*carecontext.LinkStatusEvent"},
		{"abha.care_context_discover", "*carecontext.DiscoverEvent"},
		{"abha.care_context_discover_link_init", "*carecontext.LinkInitEvent"},
		{"abha.context_discover_link_confirm", "*carecontext.LinkConfirmEvent"},
	} {
		body := fmt.Sprintf(`{"service":"abdm","event":%q,"data":{}}`, tc.event)
		e, err := ParseWebhook([]byte(body), sign(t, body, secret, time.Now()), secret)
		if err != nil {
			t.Errorf("%s: %v", tc.event, err)
			continue
		}
		if got := fmt.Sprintf("%T", e); got != tc.want {
			t.Errorf("%s -> %s, want %s", tc.event, got, tc.want)
		}
		if e.EventName() != tc.event {
			t.Errorf("EventName() = %q, want %q", e.EventName(), tc.event)
		}
	}
}

func TestParseWebhookUnknownEvent(t *testing.T) {
	body := `{"service":"abdm","event":"abha.something_new","data":{}}`
	_, err := ParseWebhook([]byte(body), sign(t, body, secret, time.Now()), secret)
	if !errors.Is(err, ErrUnknownEvent) {
		t.Errorf("err = %v, want ErrUnknownEvent", err)
	}
}

func TestParseWebhookDataFetchKeyMaterial(t *testing.T) {
	body := `{"service":"abdm","event":"abha.hip_data_fetch","data":{
      "transaction_id":"txn-1","abha_address":"p@sbx","oid":"o-1",
      "partner_patient_id":"pp-1","hip_id":"hip-1",
      "care_contexts":["cc-1"],"hi_types":["OPConsultation"],
      "key_information":{"crypto_alg":"ECDH","curve":"Curve25519",
        "dh_public_key":{"key_value":"k","parameters":"p","expiry":"e"},"nonce":"n"}}}`

	e, err := ParseWebhook([]byte(body), sign(t, body, secret, time.Now()), secret)
	if err != nil {
		t.Fatal(err)
	}
	df := e.(*DataFetchEvent)
	if df.TransactionID != "txn-1" || df.HIPID != "hip-1" || len(df.CareContexts) != 1 {
		t.Errorf("event = %+v", df)
	}
	if df.keyInfo.Nonce != "n" || df.keyInfo.DHPublicKey.KeyValue != "k" {
		t.Errorf("key material not captured: %+v", df.keyInfo)
	}
}

// An unconfigured secret must not authenticate anything: with an empty key the
// HMAC is computable by anyone, so a forged payload would otherwise pass.
func TestParseWebhookEmptySecretRejected(t *testing.T) {
	body := `{"service":"abdm","event":"abha.link_care_context","data":{"care_context_id":"cc-1"}}`
	// Signature is genuinely valid for the empty key — still must be rejected.
	_, err := ParseWebhook([]byte(body), sign(t, body, "", time.Now()), "")
	if !errors.Is(err, ErrBadSignature) {
		t.Fatalf("err = %v, want ErrBadSignature", err)
	}
}

func TestParseWebhookDataFetchEmptyData(t *testing.T) {
	body := `{"service":"abdm","event":"abha.hip_data_fetch","transaction_id":"txn-9"}`
	e, err := ParseWebhook([]byte(body), sign(t, body, secret, time.Now()), secret)
	if err != nil {
		t.Fatalf("ParseWebhook: %v", err)
	}
	if df := e.(*DataFetchEvent); df.TransactionID != "txn-9" {
		t.Errorf("transaction id = %q", df.TransactionID)
	}
}

File: java/src/test/java/care/eka/abdm/carecontext/WebhookTest.java
package care.eka.abdm.carecontext;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import org.junit.jupiter.api.Test;

class WebhookTest {

    private static final String SECRET = "whsec";

    private static String sign(String body, String secret, long ts) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            mac.update(Long.toString(ts).getBytes(StandardCharsets.UTF_8));
            mac.update((byte) '.');
            mac.update(body.getBytes(StandardCharsets.UTF_8));
            byte[] sum = mac.doFinal();
            StringBuilder sb = new StringBuilder();
            for (byte b : sum) {
                sb.append(String.format("%02x", b));
            }
            return "t=" + ts + ",v1=" + sb;
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new RuntimeException(e);
        }
    }

    private static long now() {
        return System.currentTimeMillis() / 1000;
    }

    // --- Signature verification -------------------------------------------------

    @Test
    void validSignatureDecodesLinkStatusEvent() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\","
                + "\"data\":{\"care_context_id\":\"cc-1\",\"status\":\"LINKED\"}}";
        Event e = Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sign(body, SECRET, now()), SECRET);
        LinkStatusEvent event = assertInstanceOf(LinkStatusEvent.class, e);
        assertEquals("cc-1", event.careContextId());
        assertEquals("LINKED", event.status());
    }

    @Test
    void tamperedBodyRejected() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\",\"data\":{}}";
        String sig = sign(body, SECRET, now());
        byte[] tampered = (body + " ").getBytes(StandardCharsets.UTF_8);
        assertThrows(BadSignatureException.class, () -> Webhook.parseWebhook(tampered, sig, SECRET));
    }

    @Test
    void wrongSecretRejected() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\",\"data\":{}}";
        String sig = sign(body, "other-secret", now());
        assertThrows(BadSignatureException.class,
                () -> Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sig, SECRET));
    }

    @Test
    void emptySecretRejected() {
        // javax.crypto refuses to even construct a zero-length HMAC key, so
        // there is no "genuinely valid signature for the empty key" to
        // build here (unlike Go/Python, whose HMAC implementations accept
        // one) — the empty-secret guard must reject before any HMAC is
        // computed, which this exercises with an arbitrary v1 value.
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\",\"data\":{}}";
        String header = "t=" + now() + ",v1=" + "0".repeat(64);
        assertThrows(BadSignatureException.class,
                () -> Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), header, ""));
    }

    @Test
    void staleTimestampInThePastRejected() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\",\"data\":{}}";
        long old = now() - 10 * 60;
        String sig = sign(body, SECRET, old);
        assertThrows(StaleTimestampException.class,
                () -> Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sig, SECRET));
    }

    @Test
    void staleTimestampInTheFutureAlsoRejected() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\",\"data\":{}}";
        long future = now() + 10 * 60;
        String sig = sign(body, SECRET, future);
        assertThrows(StaleTimestampException.class,
                () -> Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sig, SECRET));
    }

    @Test
    void malformedSignatureHeaderRejected() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.link_care_context\",\"data\":{}}";
        assertThrows(BadSignatureException.class,
                () -> Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), "garbage", SECRET));
    }

    // --- Event names --------------------------------------------------------------

    @Test
    void eventNameConstantsAreExactIncludingTheMissingCarePrefix() {
        assertEquals("abha.hip_data_fetch", Webhook.EVENT_DATA_FETCH);
        assertEquals("abha.link_care_context", Webhook.EVENT_LINK_STATUS);
        assertEquals("abha.care_context_discover", Webhook.EVENT_DISCOVER);
        assertEquals("abha.care_context_discover_link_init", Webhook.EVENT_LINK_INIT);
        // No care_ prefix on this one — that is not a typo.
        assertEquals("abha.context_discover_link_confirm", Webhook.EVENT_LINK_CONFIRM);
    }

    @Test
    void eachEventDecodesToTheRightType() {
        Object[][] cases = {
                {Webhook.EVENT_DATA_FETCH, DataFetchEvent.class},
                {Webhook.EVENT_LINK_STATUS, LinkStatusEvent.class},
                {Webhook.EVENT_DISCOVER, DiscoverEvent.class},
                {Webhook.EVENT_LINK_INIT, LinkInitEvent.class},
                {Webhook.EVENT_LINK_CONFIRM, LinkConfirmEvent.class},
        };
        for (Object[] c : cases) {
            String eventName = (String) c[0];
            Class<?> wantType = (Class<?>) c[1];
            String body = "{\"service\":\"abdm\",\"event\":\"" + eventName + "\",\"data\":{}}";
            Event e = Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sign(body, SECRET, now()), SECRET);
            assertInstanceOf(wantType, e, eventName);
            assertEquals(eventName, e.eventName(), eventName);
        }
    }

    @Test
    void unknownEventDistinguishableFromInauthentic() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.something_new\",\"data\":{}}";
        assertThrows(UnknownEventException.class,
                () -> Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sign(body, SECRET, now()), SECRET));
    }

    // --- Field-level details ---------------------------------------------------

    @Test
    void dataFetchCapturesKeyMaterialAndFallsBackToEnvelopeTransactionId() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.hip_data_fetch\","
                + "\"data\":{\"transaction_id\":\"txn-1\",\"abha_address\":\"p@sbx\",\"oid\":\"o-1\","
                + "\"partner_patient_id\":\"pp-1\",\"hip_id\":\"hip-1\",\"care_contexts\":[\"cc-1\"],"
                + "\"hi_types\":[\"OPConsultation\"],\"key_information\":{\"crypto_alg\":\"ECDH\","
                + "\"curve\":\"Curve25519\",\"dh_public_key\":{\"key_value\":\"k\",\"parameters\":\"p\","
                + "\"expiry\":\"e\"},\"nonce\":\"n\"}}}";
        Event e = Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sign(body, SECRET, now()), SECRET);
        DataFetchEvent event = assertInstanceOf(DataFetchEvent.class, e);
        assertEquals("txn-1", event.transactionId());
        assertEquals("hip-1", event.hipId());
        assertEquals("cc-1", event.careContexts().get(0));

        String fallbackBody = "{\"service\":\"abdm\",\"event\":\"abha.hip_data_fetch\",\"transaction_id\":\"txn-9\"}";
        Event fallback = Webhook.parseWebhook(
                fallbackBody.getBytes(StandardCharsets.UTF_8), sign(fallbackBody, SECRET, now()), SECRET);
        assertEquals("txn-9", ((DataFetchEvent) fallback).transactionId());
    }

    @Test
    void linkConfirmDecodesCamelCaseLinkRefNumber() {
        String body = "{\"service\":\"abdm\",\"event\":\"abha.context_discover_link_confirm\","
                + "\"data\":{\"linkRefNumber\":\"ref-123\",\"token\":\"111111\",\"request_id\":\"req-3\","
                + "\"oid\":\"o-1\",\"partner_patient_id\":\"pp-1\",\"hip_id\":\"hip-1\"}}";
        Event e = Webhook.parseWebhook(body.getBytes(StandardCharsets.UTF_8), sign(body, SECRET, now()), SECRET);
        LinkConfirmEvent event = assertInstanceOf(LinkConfirmEvent.class, e);
        assertEquals("ref-123", event.linkRefNumber());
        assertEquals("111111", event.token());
    }
}
